test/src/main/java/org/apache/accumulo/test/functional/MemoryFreeingIterator.java

```diff
@@ -28,23 +28,39 @@
 import org.apache.accumulo.core.iterators.IteratorEnvironment;
 import org.apache.accumulo.core.iterators.SortedKeyValueIterator;
 import org.apache.accumulo.core.iterators.WrappingIterator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
 
 public class MemoryFreeingIterator extends WrappingIterator {
 
-  @Override
-  public void init(SortedKeyValueIterator<Key,Value> source, Map<String,String> options,
-      IteratorEnvironment env) throws IOException {
-    super.init(source, options, env);
+  private static final Logger LOG = LoggerFactory.getLogger(MemoryFreeingIterator.class);
+
+  @SuppressFBWarnings(value = "DM_GC", justification = "gc is okay for test")
+  public MemoryFreeingIterator() {
+    super();
+    LOG.info("Freeing consumed memory");
     MemoryConsumingIterator.freeBuffers();
     while (this.isRunningLowOnMemory()) {
+      System.gc();
       // wait for LowMemoryDetector to recognize the memory is free.
       try {
         Thread.sleep(SECONDS.toMillis(1));
       } catch (InterruptedException ex) {
         Thread.currentThread().interrupt();
-        throw new IOException("wait for low memory detector interrupted", ex);
+        throw new RuntimeException("wait for low memory detector interrupted", ex);
       }
     }
+    LOG.info("Consumed memory freed");
+  }
+
+  @Override
+  @SuppressFBWarnings(value = "DM_GC", justification = "gc is okay for test")
+  public void init(SortedKeyValueIterator<Key,Value> source, Map<String,String> options,
+      IteratorEnvironment env) throws IOException {
+    super.init(source, options, env);
+    LOG.info("init call - should free consumed memory");
```

Comment:
What's the reasoning for moving the content of this method into the constructor? If you do need to move stuff to the constructor, you probably don't need this method at all.

Even if you keep it for some kind of log message, you don't need the warnings suppression... there is no garbage collection happening in this method anymore. Also, the log message itself seems incorrect, since it's not freeing anything.

Reply:
The reason for moving it is that the scan was not running because the server was low on memory. Instead, I moved the code to the constructor and then call `client.instanceOperations().testClassLoad(MemoryFreeingIterator.class.getName(),  WrappingIterator.class.getCanonicalName());` to load the class on the TabletServer and free the consumed memory. I don't think this method is needed.

Reply:
My only concern is that it still may not run... because calling `System.gc()` may do nothing. Is that going to result in a flaky test, or just flaky coverage but consistently passing test?

Reply:
The init() method was removed in 55265e8199.  The loop in the constructor is testing for the low memory detector to clear (show available memory). So, if the gc does not run, the the test will eventually time-out and fail, or the gc will decide that it should run either naturally, or after repeated hint requests it decides we really, really mean it.

Over-all, these tests are extremely slow to run.  But it is also a difficult situation to force and then test for recovery.  Maybe we don't run these on a regular basis?  Maybe run a subset unless specifically requested?  It does seem that we should have some coverage of the low memory detector, but I do not know what the balance is.

The low memory detector is broke right now.  Adding these changes should fix that.  We can reevaluate the tests after this is committed, but these changes do move the ball forward.

Reply:
> So, if the gc does not run, the the test will eventually time-out and fail

Resulting in a flaky test. That's my concern.

> Maybe we don't run these on a regular basis?

Running with the full ITs is fine. They shouldn't be running with sunny, though.

> We can reevaluate the tests after this is committed, but these changes do move the ball forward.

Absolutely agree :+1:  If the test is flaky, we'll deal with it then.